lib/request-proxy-send.js

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var numOrDefault = require('./util.js').numOrDefault;
+var rawHead = require('./request-proxy-util.js').rawHead;
+var strHead = require('./request-proxy-util.js').strHead;
+var TypedError = require('error/typed');
+
+var MaxRetriesExceeded = TypedError({
+    type: 'ringpop.request-proxy.max-retries-exceeded',
+    message: 'Max number of retries exceeded. ({maxRetries}) were attempted',
+    maxRetries: null
+});
+
+var RetryDivergenceError = TypedError({
+    type: 'ringpop.request-proxy.keys-diverged',
+    message: 'Destinations for proxied request have diverged. These keys: \n' +
+        ' {keys}, were originally intended for {origDestination}, but are\n' +
+        ' now meant for {newDestinations}.',
+    keys: null,
+    origDestination: null,
+    newDestinations: null
+});
+
+var RETRY_SCHEDULE = [0, 1, 3.5];
+
+function RequestProxySend(opts) {
+    this.ringpop = opts.ringpop;
```

Comment:
nitpick: I think this pattern of passing the entire `ringpop` interface into the Send operation is not a good pattern.

It would be great if we could pass in just the interface that `RequestProxySend` actually needs.

`{ stat, logger, ring: { whoami, lookup }, emit, channel }`

This way we can see what it actually needs to operate. A `channel` to send() to and a ringpop interface for whoami() and lookup()